File: web/src/lib/scanners.ts
// How each scanner is described in the app, and how its evidence is shown.
// The wording comes from the TechnoFunda methodology: these are the price-volume
// screens that start the funnel.

import { crore, money, percent, times } from "./format";

export type Evidence = Record<string, string | number | null>;

export type ScannerInfo = {
  name: string;
  what: string;
  /** Turns stored evidence into "why it triggered", in plain words. */
  explain: (e: Evidence) => { label: string; value: string }[];
};

const n = (e: Evidence, key: string): number | null => {
  const value = e[key];
  return typeof value === "number" ? value : null;
};

export const SCANNERS: Record<string, ScannerInfo> = {
  consolidation_breakout: {
    name: "Consolidation breakout",
    what: "Price left a long sideways range on heavy volume",
    explain: (e) => [
      { label: "Broke above", value: money(n(e, "breakout_level")) },
      { label: "Close", value: money(n(e, "close")) },
      { label: "Above level by", value: percent(n(e, "above_level_pct")) },
      { label: "Days in range", value: `${n(e, "days_in_range") ?? "—"}` },
      { label: "Range width", value: percent(n(e, "range_width")) },
      { label: "Volume", value: `${times(n(e, "volume_ratio_20"))} the 20-day average` },
    ],
  },
  new_high_breakout: {
    name: "New high breakout",
    what: "Price reached a 52-week high while the long-term trend is rising",
    explain: (e) => [
      { label: "Close", value: money(n(e, "close")) },
      { label: "52-week high", value: money(n(e, "high_52w")) },
      { label: "Below the high", value: percent(n(e, "pct_from_high_52w"), 2) },
      { label: "Volume", value: `${times(n(e, "volume_ratio_20"))} the 20-day average` },
      { label: "30-week average", value: money(n(e, "wma_30w")) },
      { label: "30-week trend", value: `${percent(n(e, "wma_30w_slope"))} over 10 weeks` },
    ],
  },
  volume_expansion: {
    name: "Volume expansion",
    what: "Volume far above its own average, with price not falling",
    explain: (e) => [
      { label: "Close", value: money(n(e, "close")) },
      { label: "Volume", value: `${times(n(e, "volume_ratio_20"))} the 20-day average` },
      {
        label: "20-day average volume",
        value: (n(e, "volume_avg_20") ?? 0).toLocaleString("en-IN", { maximumFractionDigits: 0 }),
      },
      { label: "Day's move", value: percent(n(e, "return_1d")) },
      { label: "Below 52-week high", value: percent(n(e, "pct_from_high_52w")) },
    ],
  },
  price_volume_surge: {
    name: "Price + volume surge",
    what: "An unusual price move on unusual volume — the methodology's first-level screen",
    explain: (e) => [
      { label: "Close", value: money(n(e, "close")) },
      { label: "Day's move", value: percent(n(e, "return_1d")) },
      { label: "Volume", value: `${times(n(e, "volume_ratio_20"))} the 20-day average` },
      { label: "Below 52-week high", value: percent(n(e, "pct_from_high_52w")) },
    ],
  },
  sector_trend: {
    name: "Sector trend",
    what: "A rising stock inside a sector that is beating the market",
    explain: (e) => [
      { label: "Close", value: money(n(e, "close")) },
      { label: "Sector index", value: `${e.sector_index ?? "—"}` },
      { label: "Sector, 1 month", value: percent(n(e, "sector_return_21d")) },
      { label: "Sector vs market", value: percent(n(e, "sector_relative_21d")) },
      { label: "Sector rank", value: `${(n(e, "sector_rank") ?? 0).toFixed(0)} / 100` },
      { label: "Stock strength rank", value: `${(n(e, "rs_rank_63d") ?? 0).toFixed(0)} / 100` },
    ],
  },
};

/** Accent colour token for each scanner (defined in globals.css). */
const COLOURS: Record<string, string> = {
  price_volume_surge: "scan-surge",
  volume_expansion: "scan-volume",
  consolidation_breakout: "scan-breakout",
  new_high_breakout: "scan-high",
  sector_trend: "scan-sector",
};

export function scannerColour(key: string): string {
  return COLOURS[key] ?? "scan-high";
}

/** The order the methodology applies them: price-volume first, then structure. */
export const SCANNER_ORDER = [
  "price_volume_surge",
  "volume_expansion",
  "consolidation_breakout",
  "new_high_breakout",
  "sector_trend",
];

export function scannerInfo(key: string): ScannerInfo {
  return (
    SCANNERS[key] ?? {
      name: key,
      what: "",
      explain: (e) =>
        Object.entries(e)
          .filter(([k]) => k !== "symbol")
          .map(([k, v]) => ({ label: k, value: String(v) })),
    }
  );
}

/** A short one-line summary for the results table. */
export function headline(key: string, e: Evidence): string {
  switch (key) {
    case "consolidation_breakout":
      return `Broke ${money(n(e, "breakout_level"))} after ${n(e, "days_in_range")} days in range, ${times(n(e, "volume_ratio_20"))} volume`;
    case "new_high_breakout":
      return `At its 52-week high of ${money(n(e, "high_52w"))}, ${times(n(e, "volume_ratio_20"))} volume`;
    case "volume_expansion":
      return `${times(n(e, "volume_ratio_20"))} its average volume, price ${percent(n(e, "return_1d"))}`;
    case "price_volume_surge":
      return `Up ${percent(n(e, "return_1d"))} on ${times(n(e, "volume_ratio_20"))} volume`;
    case "sector_trend":
      return `${e.sector_index ?? "Sector"} beating the market by ${percent(n(e, "sector_relative_21d"))}`;
    default:
      return "";
  }
}

export { crore };

File: jobs/scanners.py
"""The scanners: rules that pick out stocks worth looking at.

Each scanner is a plain function of one day's numbers for one company, so it can be tested
with made-up values and replayed over history. A scanner returns the evidence behind the
trigger (or None), and that evidence is stored with the signal, so the app can always answer
"why did this appear?" without anyone reading the code.

Thresholds live in config/scanners.json, not here, and every signal records which version
of that file produced it.
"""

import json
from pathlib import Path
from typing import Any

CONFIG_PATH = Path(__file__).resolve().parent.parent / "config" / "scanners.json"

Row = dict[str, Any]  # one company's features for one day, plus sector values


def load_config(path: Path = CONFIG_PATH) -> dict:
    return json.loads(path.read_text(encoding="utf-8"))


def _has(row: Row, *names: str) -> bool:
    """True when every named value exists (long windows are empty early in a stock's history)."""
    return all(row.get(n) is not None for n in names)


def is_liquid(row: Row, config: dict) -> bool:
    limit = config["liquidity"]["min_value_avg_20"]
    return _has(row, "value_avg_20") and row["value_avg_20"] >= limit


# ---------------------------------------------------------------------------
# Scanners
# ---------------------------------------------------------------------------
def consolidation_breakout(row: Row, config: dict) -> dict | None:
    """Price leaves a long sideways range (the methodology's "bamboo" setup).

    `days_in_range_prev` is yesterday's count, because today's breakout resets it to 0.
    """
    c = config["consolidation_breakout"]
    if not _has(row, "close_adj", "range_high_120", "range_width_120", "volume_ratio_20"):
        return None
    if row.get("days_in_range_prev") is None:
        return None
    if row["close_adj"] <= row["range_high_120"]:
        return None
    if row["days_in_range_prev"] < c["min_days_in_range"]:
        return None
    if row["range_width_120"] > c["max_range_width"]:
        return None
    if row["volume_ratio_20"] < c["min_volume_ratio"]:
        return None
    return {
        "close": row["close_adj"],
        "breakout_level": row["range_high_120"],
        "above_level_pct": (row["close_adj"] - row["range_high_120"]) / row["range_high_120"],
        "days_in_range": row["days_in_range_prev"],
        "range_width": row["range_width_120"],
        "volume_ratio_20": row["volume_ratio_20"],
        "pct_from_high_52w": row.get("pct_from_high_52w"),
    }


def new_high_breakout(row: Row, config: dict) -> dict | None:
    """Price reaches a new 52-week high with the long-term trend rising."""
    c = config["new_high_breakout"]
    if not _has(row, "close_adj", "high_52w", "pct_from_high_52w", "volume_ratio_20"):
        return None
    if row["pct_from_high_52w"] > c["max_pct_from_high_52w"]:
        return None
    if row["volume_ratio_20"] < c["min_volume_ratio"]:
        return None
    if c["require_rising_30w"]:
        if not _has(row, "wma_30w_slope") or row["wma_30w_slope"] <= 0:
            return None
    return {
        "close": row["close_adj"],
        "high_52w": row["high_52w"],
        "pct_from_high_52w": row["pct_from_high_52w"],
        "volume_ratio_20": row["volume_ratio_20"],
        "wma_30w": row.get("wma_30w"),
        "wma_30w_slope": row.get("wma_30w_slope"),
        "rs_rank_63d": row.get("rs_rank_63d"),
    }


def volume_expansion(row: Row, config: dict) -> dict | None:
    """Volume far above its own average: someone is doing something."""
    c = config["volume_expansion"]
    if not _has(row, "volume_ratio_20", "return_1d", "volume_avg_20"):
        return None
    if row["volume_ratio_20"] < c["min_volume_ratio"]:
        return None
    if row["return_1d"] < c["min_return_1d"]:
        return None
    return {
        "close": row.get("close_adj"),
        "volume_ratio_20": row["volume_ratio_20"],
        "volume_avg_20": row["volume_avg_20"],
        "return_1d": row["return_1d"],
        "pct_from_high_52w": row.get("pct_from_high_52w"),
    }


def price_volume_surge(row: Row, config: dict) -> dict | None:
    """Unusual price move on unusual volume — the methodology's first-level screen."""
    c = config["price_volume_surge"]
    if not _has(row, "return_1d", "volume_ratio_20"):
        return None
    if row["return_1d"] < c["min_return_1d"]:
        return None
    if row["volume_ratio_20"] < c["min_volume_ratio"]:
        return None
    return {
        "close": row.get("close_adj"),
        "return_1d": row["return_1d"],
        "volume_ratio_20": row["volume_ratio_20"],
        "pct_from_high_52w": row.get("pct_from_high_52w"),
        "sector": row.get("sector"),
    }


def sector_trend(row: Row, config: dict) -> dict | None:
    """A rising stock whose sector has just taken the lead.

    This is a state that can last for weeks, so it only triggers on the day it becomes
    true - when the sector moves into the leading group, or the stock climbs back above
    its 30-week average. Otherwise every stock in a strong sector would signal every day.
    """
    c = config["sector_trend"]
    if not _has(row, "sector_rank_relative_21d", "sector_relative_21d", "rs_rank_63d"):
        return None
    if row["sector_rank_relative_21d"] < c["min_sector_rank"]:
        return None
    if row["sector_relative_21d"] < c["min_relative_21d"]:
        return None
    if row["rs_rank_63d"] < c["min_rs_rank_63d"]:
        return None
    if c["require_above_30w"]:
        if not _has(row, "close_adj", "wma_30w") or row["close_adj"] <= row["wma_30w"]:
            return None

    # Only on the day it starts: yesterday the sector was not leading, or the stock
    # was not above its 30-week average.
    sector_was_leading = (row.get("sector_rank_relative_21d_prev") or 0) >= c["min_sector_rank"]
    was_above_30w = (row.get("close_adj_prev") or 0) > (row.get("wma_30w_prev") or 0)
    if sector_was_leading and was_above_30w:
        return None

    return {
        "close": row.get("close_adj"),
        "started_because": "sector moved into the lead"
        if not sector_was_leading
        else "price climbed back above its 30-week average",
        "sector": row.get("sector"),
        "sector_index": row.get("sector_index"),
        "sector_return_21d": row.get("sector_return_21d"),
        "sector_relative_21d": row["sector_relative_21d"],
        "sector_rank": row["sector_rank_relative_21d"],
        "rs_rank_63d": row["rs_rank_63d"],
        "wma_30w": row.get("wma_30w"),
    }


SCANNERS = {
    "consolidation_breakout": consolidation_breakout,
    "new_high_breakout": new_high_breakout,
    "volume_expansion": volume_expansion,
    "price_volume_surge": price_volume_surge,
    "sector_trend": sector_trend,
}


def run_all(row: Row, config: dict) -> dict[str, dict]:
    """Run every scanner on one company-day. Returns {scanner name: evidence}."""
    if not is_liquid(row, config):
        return {}
    found = {}
    for name, scanner in SCANNERS.items():
        evidence = scanner(row, config)
        if evidence is not None:
            found[name] = evidence
    return found
